File: java/src/main/java/jline/lang/layered/LayeredNetworkStruct.java
package jline.lang.layered;

import jline.lang.constant.CallType;
import jline.lang.constant.SchedStrategy;
import jline.lang.distributions.DiscreteDistribution;
import jline.lang.distributions.Distribution;
import jline.lang.distributions.Geometric;
import jline.util.Matrix;

import java.util.List;
import java.util.Map;

/**
 * Class summarizing the characteristics of a LayeredNetwork
 */
public class LayeredNetworkStruct {
    public int nidx;
    public int nhosts;
    public int ntasks;
    public int nacts;
    public int nentries;
    public int ncalls;

    public int tshift;
    public int eshift;
    public int ashift;
    public int hshift;
    public int cshift;

    public Map<Integer, List<Integer>> tasksof;
    public Map<Integer, List<Integer>> entriesof;
    public Map<Integer, List<Integer>> actsof;
    public Map<Integer, List<Integer>> callsof;
    public Map<Integer, Distribution> hostdem;
    public Map<Integer, Distribution> think;
    public Map<Integer, Integer> itemcap;
    public Map<Integer, DiscreteDistribution> itemproc;
    public Map<Integer, CallType> calltype;
    public Map<Integer, Geometric> callproc;
    public Map<Integer, String> callnames;
    public Map<Integer, String> callhashnames;

    public Map<Integer, SchedStrategy> sched;
    public Matrix schedid;
    public Map<Integer, String> names;
    public Map<Integer, String> hashnames;
    public Matrix mult;
    public Matrix repl;
    public Matrix type;
    public Matrix graph;

    public Matrix replygraph;
    public Matrix nitems;

    public Matrix replacement;
    public Matrix iscache;
    public Matrix parent;

    public Matrix iscaller;
    public Matrix issynccaller;
    public Matrix isasynccaller;
    public Matrix callpair;
    public Matrix taskgraph;

    public Matrix actpretype;
    public Matrix actposttype;

    public Matrix isref;

    public Matrix conntasks; // 1 * n matrix

    public LayeredNetworkStruct () {

    }

    public void print() {
        System.out.println("nidx: " + nidx);
        System.out.println("nhosts: " + nhosts);
        System.out.println("ntasks: " + ntasks);
        System.out.println("nentries: " + nentries);
        System.out.println("nacts: " + nacts);
        System.out.println("ncalls: " + ncalls);
        System.out.println("hshift: " + hshift);
        System.out.println("tshift: " + tshift);
        System.out.println("eshift: " + eshift);
        System.out.println("ashift: " + ashift);
        System.out.println("cshift: " + cshift);
        System.out.println("tasksof: " + tasksof);
        System.out.println("entriesof: " + entriesof);
        System.out.println("actof: " + actsof);
        System.out.println("callof: " + callsof);
        System.out.println("hostdem: " + hostdem);
        System.out.println("think: " + think);
        System.out.println("sched: " + sched);
        System.out.println("schedid: ");
        schedid.print();
        System.out.println("names: " + names);
        System.out.println("hashnames: " + hashnames);
        System.out.println("mult: ");
        mult.print();
        System.out.println("repl: ");
        repl.print();
        System.out.println("type: ");
        type.print();
        System.out.println("nitems: ");
        nitems.print();
        System.out.println("itemcap: " + itemcap);
        System.out.println("replacement: ");
        replacement.print();
        System.out.println("itemproc: " + itemproc);
        System.out.println("calltype: " + calltype);

        System.out.println("callpair: ");
        callpair.print();
        System.out.println("callproc: " + callproc);

        System.out.println("callnames: " + callnames);
        System.out.println("callhashname: " + callhashnames);
        System.out.println("actpretype: ");
        actpretype.print();
        System.out.println("actposttype: ");
        actposttype.print();
        System.out.println("graph: ");
        graph.printNonZero();
        System.out.println("parent: ");
        parent.print();
        System.out.println("replygraph: ");
        replygraph.printNonZero();
        System.out.println("iscache: ");
        iscache.print();
        System.out.println("iscaller: ");
        iscaller.printNonZero();
        System.out.println("issynccaller: ");
        issynccaller.printNonZero();
        System.out.println("isasynccaller: ");
        isasynccaller.printNonZero();
        System.out.println("isref: ");
        isref.print();
    }

}

File: matlab/src/lang/layered/LayeredNetworkStruct.m
function lqn=LayeredNetworkStruct()
% Data structure representation for a LayeredNetwork object
%
% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

lqn=[]; %faster than lqn=struct();
lqn.nidx = 0;  % total number of hosts, tasks, entries, and activities, except the reference tasks
lqn.nhosts = 0;
lqn.ntasks = 0;
lqn.nentries = 0;
lqn.nacts = 0;
lqn.ncalls = 0;
lqn.hshift = 0;
lqn.tshift = 0;
lqn.eshift = 0;
lqn.ashift = 0;
lqn.cshift = 0;
lqn.nidx = 0;
lqn.tasksof = {};
lqn.entriesof = {};
lqn.actsof = {};
lqn.callsof = {};
lqn.hostdem = {};
lqn.hostdem_type = [];      % ProcessType enum vector
lqn.hostdem_params = {};    % Cell array of parameter vectors
lqn.hostdem_mean = [];      % Precomputed mean vector
lqn.hostdem_scv = [];       % Precomputed SCV vector
lqn.hostdem_proc = {};      % MAP/PH process {D0,D1} matrices

lqn.actthink = {};
lqn.actthink_type = [];
lqn.actthink_params = {};
lqn.actthink_mean = [];
lqn.actthink_scv = [];
lqn.actthink_proc = {};

lqn.think = {};
lqn.think_type = [];
lqn.think_params = {};
lqn.think_mean = [];
lqn.think_scv = [];
lqn.think_proc = {};
lqn.sched = [];
lqn.names = {};
lqn.hashnames = {};
%lqn.shortnames = {};
lqn.mult = [];
lqn.repl = [];
% Task scheduling priority, 0 on every other index. Lower is served first,
% the convention sn.classprio uses in a Network. Read by HOL scheduling.
lqn.prio = [];
lqn.type = [];
%lqn.replies = [];
lqn.parent = [];

lqn.nitems = [];
lqn.itemcap  = {};
lqn.replacestrat  = [];

lqn.itemproc = {};
lqn.itemproc_type = [];
lqn.itemproc_params = {};
lqn.itemproc_mean = [];
lqn.itemproc_scv = [];
lqn.itemproc_proc = {};

lqn.setuptime = {};
lqn.setuptime_type = [];
lqn.setuptime_params = {};
lqn.setuptime_mean = [];
lqn.setuptime_scv = [];
lqn.setuptime_proc = {};

lqn.delayofftime = {};
lqn.delayofftime_type = [];
lqn.delayofftime_params = {};
lqn.delayofftime_mean = [];
lqn.delayofftime_scv = [];
lqn.delayofftime_proc = {};
lqn.calltype = sparse([]);
lqn.callpair = [];

lqn.callproc = {};
lqn.callproc_type = [];
lqn.callproc_params = {};
lqn.callproc_mean = [];
lqn.callproc_scv = [];
lqn.callproc_proc = {};

lqn.callnames = {};
lqn.callhashnames = {};

% Open arrival distributions (populated in getStruct for entries with arrivals)
lqn.arrival = {};
lqn.arrival_type = [];
lqn.arrival_params = {};
lqn.arrival_mean = [];
lqn.arrival_scv = [];
lqn.arrival_proc = {};
%lqn.callshortnames = {};
lqn.actpretype = sparse([]);
lqn.actposttype = sparse([]);

lqn.graph = sparse([]);
lqn.taskgraph = sparse([]);
lqn.replygraph = [];
lqn.actphase = [];  % Phase number (1..3) for each activity

lqn.lincon = {};            % cell(nhosts+ntasks,2); lincon{i,1}=Matrix(C_i,K_i), lincon{i,2}=Matrix(C_i,1): admission A*n<=b on host/task i's layer station. Empty if none

lqn.lldscaling = {};        % cell(nhosts+ntasks,1); lldscaling{i}=vector alpha(n) scaling host/task i's layer-station rate at total population n. Empty where absent
lqn.cdscaling = {};         % cell(nhosts+ntasks,1); cdscaling{i}=product-form handle beta(n) over that server's operands, tasksof/entriesof order as lincon. Empty where absent
lqn.cdscalingpeak = {};     % cell(nhosts+ntasks,1); per-operand peak rate scaling of cdscaling{i}, for the Util=T*S/peak normalization
lqn.jdscaling = {};         % cell(nhosts+ntasks,1); non-product-form handle eta(n) over that server's operands. Empty where absent
lqn.jdscalingpeak = {};     % cell(nhosts+ntasks,1); per-operand peak rate scaling of jdscaling{i}
lqn.pools = {};             % cell(nhosts+ntasks,1); struct('names','counts','rates','compat') of the compatibility pools declared on that server. compat(t,j) nonzero = pool t may serve operand j. Empty where absent; see sn_compat_rate

lqn.iscache = sparse(logical([]));
lqn.iscaller = sparse([]);
lqn.issynccaller = sparse([]);
lqn.isasynccaller = sparse([]);
lqn.isref = sparse(logical([]));
lqn.hassetup = sparse(logical([]));
end
